compiler/next/tools/chpldoc/chpldoc.cpp

```diff
@@ -0,0 +1,128 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <algorithm>
+
+#include "chpl/parsing/Parser.h"
+#include "chpl/queries/Context.h"
+#include "chpl/uast/ASTTypes.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Comment.h"
+
+using namespace chpl;
+using namespace uast;
+using namespace parsing;
+
+// Remove the leading+trailing // or /* (*/)
+std::string stripComment(const std::string &s) {
+  if (s.substr(0, 2) == "//") return s.substr(2);
+  if (s.substr(0, 2) == "/*") {
+    size_t l = s.length();
+    assert(s.substr(l - 2 - 1, 2) == "*/");
+    return s.substr(2, l - 4);
+  }
+  assert(false);
+  return "";
+}
+
+const char *testString = "/* comment 1 */\n"
+                         "module M {\n"
+                         "  // comment 2 \n"
+                         "  x;\n"
+                         "  /* comment 3 */\n"
+                         "}\n"
+                         "/* comment 4 */";
+
+int main(int argc, char **argv) {
+  Context context;
+  Context* ctx = &context;
+
+  auto parser = Parser::build(ctx);
+  Parser* p = parser.get();
+
+  auto parseResult = argc == 2 ?
+    parser->parseFile(argv[1]) :
+    parser->parseString("test1a.chpl", testString);
+
+  if (parseResult.numErrors()) {
+    // for (auto &err : parseResult.errors)
+    for (size_t i = 0; i < parseResult.numErrors(); i++) {
+      // std::cerr << parseResult.error(i) << '\n';
+      auto &err = parseResult.error(i);
+      std::cerr << err.path().c_str() << ":" << err.line() << ":" << err.firstColumn() << ":" << err.message() << "\n";
+    }
+    return 1;
+  }
+
+  std::cerr << "got " << parseResult.numTopLevelExpressions() << " top level expressions\n";
+
+  // TODO why can't I do this
+  // const auto &topLevelExpressions = parseResult.topLevelExpressions();
+  // auto it = std::find_if(topLevelExpressions.begin(), topLevelExpressions.end(), [](const auto &x){ return false; });
+
+  int module_i = -1;
+  for (ssize_t i = 0; i < parseResult.numTopLevelExpressions(); i++) {
+    if (parseResult.topLevelExpression(i)->isModule()) {
+      module_i = i;
+      break;
+    }
+  }
+
+  if (module_i == -1) {
+    std::cerr << "Did not find any top level module expressions\n";
+    return 1;
+  }
+
+  // Print module title
+  const Module *mod = parseResult.topLevelExpression(module_i)->toModule();
+  std::cout << mod->name().c_str() << "\n";
+  std::cout << std::string(mod->name().length(), '=') << "\n\n";
+
+  // Print preceding comment to module
+  if (module_i >= 1) {
+    if (const Comment *comment = parseResult.topLevelExpression(module_i - 1)->toComment()) {
```

Comment:
@mppf Are we trying to restrict line width to 80 cols in the new codebase? Or do we not care about that for drivers?

Reply:
Yes, let's try to limit to 80 column lines.